### PR TITLE
add register action request/response

### PR DESCRIPTION
**Pull request by ylwu-amzn**
### Description
Add register agent action request/response 
 
### Issues Resolved
[List any issues this PR will resolve]
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: common/src/main/java/org/opensearch/ml/common/transport/undeploy/MLUndeployModelsResponse.java [hunk 1]
```diff
@@ -49,4 +54,20 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
         }
         return builder;
     }
+
+    public static MLUndeployModelsResponse fromActionResponse(ActionResponse actionResponse) {
```
Comment (dhrubo-os) [inline common/src/main/java/org/opensearch/ml/common/transport/undeploy/MLUndeployModelsResponse.java:58]: Can we add test for this class?

File: common/src/main/java/org/opensearch/ml/common/agent/LLMSpec.java [hunk 2]
```diff
@@ -19,7 +20,7 @@
 import static org.opensearch.core.xcontent.XContentParserUtils.ensureExpectedToken;
 import static org.opensearch.ml.common.utils.StringUtils.getParameterMap;
 
-
+@EqualsAndHashCode
```
Comment (dhrubo-os) [inline common/src/main/java/org/opensearch/ml/common/agent/LLMSpec.java:23]: Just curious to know why do we need this?

File: common/src/main/java/org/opensearch/ml/common/agent/MLAgent.java [hunk 3]
```diff
@@ -99,15 +100,17 @@ public MLAgent(StreamInput input) throws IOException{
         if (input.readBoolean()) {
             memory = new MLMemorySpec(input);
         }
-        createdTime = input.readInstant();
-        lastUpdateTime = input.readInstant();
-        appType = input.readString();
+        createdTime = input.readOptionalInstant();
+        lastUpdateTime = input.readOptionalInstant();
+        appType = input.readOptionalString();
```
Comment (dhrubo-os) [inline common/src/main/java/org/opensearch/ml/common/agent/MLAgent.java:105]: isn't app type mandatory for memory? or this is something different?

Comment (ylwu-amzn) [reply common/src/main/java/org/opensearch/ml/common/agent/MLAgent.java:105]: It's different, we have memory type for memory

File: common/src/main/java/org/opensearch/ml/common/agent/LLMSpec.java (hunk 2 above)
Comment (ylwu-amzn) [reply common/src/main/java/org/opensearch/ml/common/agent/LLMSpec.java:23]: Mainly to make it easier to write UT to assert equals

File: common/src/main/java/org/opensearch/ml/common/transport/undeploy/MLUndeployModelsResponse.java (hunk 1 above)
Comment (ylwu-amzn) [reply common/src/main/java/org/opensearch/ml/common/transport/undeploy/MLUndeployModelsResponse.java:58]: sure

File: common/src/main/java/org/opensearch/ml/common/agent/MLAgent.java [hunk 4]
```diff
@@ -99,15 +100,17 @@ public MLAgent(StreamInput input) throws IOException{
...
             Set<String> toolNames = new HashSet<>();
-            for (MLToolSpec toolSpec : tools) {
-                String toolName = Optional.ofNullable(toolSpec.getName()).orElse(toolSpec.getType());
-                if (toolNames.contains(toolName)) {
-                    throw new IllegalArgumentException("Tool has duplicate name or alias: " + toolName);
+            if (tools != null) {
+                for (MLToolSpec toolSpec : tools) {
+                    String toolName = Optional.ofNullable(toolSpec.getName()).orElse(toolSpec.getType());
+                    if (toolNames.contains(toolName)) {
+                        throw new IllegalArgumentException("Tool has duplicate name or alias: " + toolName);
```
Comment (dhrubo-os) [inline common/src/main/java/org/opensearch/ml/common/agent/MLAgent.java:112]: Can we cover this section in test?.

Comment (ylwu-amzn) [reply common/src/main/java/org/opensearch/ml/common/agent/MLAgent.java]: sure

**Review by dhrubo-os — APPROVED** [review]

File: common/src/main/java/org/opensearch/ml/common/agent/MLAgent.java [hunk 5]
```diff
@@ -77,6 +75,24 @@ public MLAgent(String name,
...
+    private void validate() {
+        if (name == null) {
+            throw new IllegalArgumentException("agent name is null");
+        }
+        Set<String> toolNames = new HashSet<>();
+        if (tools != null) {
+            for (MLToolSpec toolSpec : tools) {
+                String toolName = Optional.ofNullable(toolSpec.getName()).orElse(toolSpec.getType());
+                if (toolNames.contains(toolName)) {
+                    throw new IllegalArgumentException("Duplicate tool defined: " + toolName);
```
Comment (Zhangxunmt) [inline common/src/main/java/org/opensearch/ml/common/agent/MLAgent.java:90]: Is it expected to throw this exception or just remove the redundancy in the tools and return normally?

**Review by Zhangxunmt — APPROVED** [review]

File: common/src/main/java/org/opensearch/ml/common/agent/MLAgent.java (hunk 5 above)
Comment (ylwu-amzn) [reply common/src/main/java/org/opensearch/ml/common/agent/MLAgent.java:90]: Yes, it's expected, we can't remove redundancy for the user automatically as we don't know which one is the correct.

**Pull request merged**